[blog] Add 0.73 announcement post (#3931)
facebook/react-native-website: website/blog/2023-12-06-0.73-debugging-improvements-stable-symlinks.md

Remote JavaScript Debugging is a legacy debugging mode that connects an external web browser (Chrome) to your app and runs your JavaScript code inside a web page, i.e. `http://localhost:8081/debugger-ui`. This model could lead to inconsistent app behaviour while debugging, and is incompatible with native modules under the New Architecture.

In 0.73, Remote JavaScript Debugging is deprecated and has been removed from the Dev Menu. Enabling the remote debugger must now be done manually via the `NativeDevSettings` API. Doing this is covered in the [Other Debugging Methods docs](/docs/next/other-debugging-methods#remote-js-debugging).

:::info
Remote JavaScript Debugging was previously the default debugging experience for apps using JavaScriptCore (JSC). We recommend [Safari Developer Tools (direct JSC debugging)](/docs/next/other-debugging-methods#safari-developer-tools-direct-jsc-debugging) instead, for iOS apps.

We recommend using [Hermes](/docs/hermes) for a consistent debugging experience on all platforms.
:::

### New Architecture Updates

We continue active development on the New Architecture and are fully invested in bringing it to the Open Source community.

> In 0.73, Bridgeless Mode is now available for experimentation. We're shipping this with a new interop layer to continue support for legacy Native Modules. And we're introducing Static View Configs, which simplify native view components on iOS.

Reviewer: @huntie Let me rephrase this paragraph tomorrow please 🙏

Reviewer 2: Static ViewConfigs is just an optimization. Let's remove it from this blurb. That way, we focus only on the important bits: Bridgeless mode and its backwards compatibility layers. 

People can always learn about Static ViewConfigs when they read the Bridgeless mode stable working group post.

---
title: 'React Native 0.73 - Debugging Improvements, Stable Symlink Support, and more'
authors: [fortmarek, titozzz, hurali, lunaleaps, huntie]
tags: [announcement, release, debugging]
date: 2023-12-06
---

> Today we're releasing React Native 0.73! This release adds improvements to debugging with Hermes, stable symlink support, Android 14 support, and new experimental features. We are also deprecating legacy debugging features. We continue to make foundational improvements in preparation for the New Architecture.

Reviewer 2: Bridgeless mode is a very important milestone for the new architecture rollout in open source. We should try to get more people using and adopting bridgeless mode in 0.73. So, let's not hide it under "foundational improvements" or "New Architecture updates", and highlight it from the get-go so we get peoples' eyes on it.

```suggestion
Today we're releasing React Native 0.73! This release adds improvements to debugging with Hermes, stable symlink support, Android 14 support, and new experimental features. We are also deprecating legacy debugging features. We are also releasing the next pillar of the New Architecture: Bridgeless Mode!
```

Reviewer 2: @huntie, how do you feel about moving this higher up in the list, also?

We've updated React Native's template on Android to use Kotlin instead of Java. The new `MainActivity.kt` and `MainApplication.kt` files are 36% smaller in size.

The [Upgrade Helper](https://react-native-community.github.io/upgrade-helper/) has also been updated to make it easier to migrate your `.java` files to `.kt` files.
If you've previously modified the Java files in your project and you need support migrating them to Kotlin, you can use the `Code > Convert Java file to Kotlin File` utility of Android Studio (also accessible with the shortcut <kbd>Cmd ⌘</kbd> + <kbd>Shift ⇧</kbd> + <kbd>Option ⌥</kbd> + <kbd>K</kbd>).

### Android 14 Support

We've updated React Native to fully support Android 14. Starting from 0.73, React Native developers can now target the latest Android SDK version, [API Level 34](https://developer.android.com/guide/topics/manifest/uses-sdk-element?hl=en#ApiLevels) (_Upside Down Cake_).

#### Grant partial access to photos and videos

> [Selected Photos Access](https://developer.android.com/about/versions/14/changes/partial-photo-video-access) allows Android 14 users to grant apps access to specific items in their media library, rather than access to all media. In 0.73, React Native apps now support this capability, by using the `READ_MEDIA_VISUAL_USER_SELECTED` permission in the [`PermissionsAndroid`](https://reactnative.dev/docs/permissionsandroid) API.

Reviewer 3: ```suggestion
[Selected Photos Access](https://developer.android.com/about/versions/14/changes/partial-photo-video-access) allows Android 14 users to grant apps access to specific items in their media library, rather than access to all media. In 0.73, React Native apps now support this capability, by using the `READ_MEDIA_VISUAL_USER_SELECTED` permission in the [`PermissionsAndroid`](/docs/permissionsandroid) API.
```

Internal link nit.

#### Min SDK bump

> React Native 0.73 will be the final version which supports Android 5.0 (API Level 21). The next version of React Native will have a minimum SDK version of 23 (Android 6.0). Read more about the upcoming changes to min SDK bump [here](/). <!-- TODO -->

Reviewer 4: there's a missing link here at the end

Reviewer: Yup this will be published tomorrow

Reviewer: Link is here @huntie:
- https://github.com/react-native-community/discussions-and-proposals/discussions/740

### Deprecated Debugging Features

#### Flipper ↔ React Native integration

We are gradually moving away from Flipper as the default tool for debugging React Native apps. In 0.73, this begins by deprecating the native Flipper integration included with React Native (bootstrap code which wires up Flipper's core plugins). We will be removing this integration and dependency in the next release — meaning parts of Flipper such as the Network plugin will cease to work.

> **Unchanged**: Flipper as a standalone product for native app debugging will continue to exist. Even after its removal from new React Native projects in future, developers will be able to [manually add Flipper to their app](https://fbflipper.com/docs/getting-started/android-native/) if they wish.

Reviewer 3: ```suggestion
**Unchanged**: Flipper as a standalone product for native app debugging will continue to exist. Even after its removal from new React Native projects in the future, developers will be able to [manually add Flipper to their app](https://fbflipper.com/docs/getting-started/android-native/) if they wish.
```

Gramma nit.

#### Remote JavaScript Debugging

Remote JavaScript Debugging is a legacy debugging mode that connects an external web browser (Chrome) to your app and runs your JavaScript code inside a web page, i.e. `http://localhost:8081/debugger-ui`. This model could lead to inconsistent app behaviour while debugging, and is incompatible with native modules under the New Architecture.

In 0.73, Remote JavaScript Debugging is deprecated and has been removed from the Dev Menu. Enabling the remote debugger must now be done manually via the `NativeDevSettings` API. Doing this is covered in the [Other Debugging Methods docs](/docs/next/other-debugging-methods#remote-js-debugging).

:::info
Remote JavaScript Debugging was previously the default debugging experience for apps using JavaScriptCore (JSC). We recommend [Safari Developer Tools (direct JSC debugging)](/docs/next/other-debugging-methods#safari-developer-tools-direct-jsc-debugging) instead, for iOS apps.

We recommend using [Hermes](/docs/hermes) for a consistent debugging experience on all platforms.
:::

### New Architecture Updates

> We continue active development on the New Architecture and are fully invested in bringing it to the Open Source community.
> 
> In 0.73, Bridgeless Mode is now available for experimentation. We're shipping this with a new interop layer to continue support for legacy Native Modules. And we're introducing Static View Configs, which simplify native view components on iOS.
> 
> More info on these features will be covered in an incoming blog post.

Reviewer: ```suggestion
We continue the rollout of the New Architecture of React Native, to make it available to everyone in the Open Source community.

Since React Native 0.68, both the New Renderer (Fabric) and the New Native Module System (TurboModule) were available to users to experiment and evaluate. We want to thank the community for the feedback we received so far.

Today we're releasing another piece of the New Architecture: **Bridgeless mode**. Up until now, when you enable the New Architecture is your apps, the Bridge was still available to support backward compatibility with older components and modules. However, our vision is to fully sunset the bridge. Starting from React Native 0.73, you can enable Bridgeless Mode which will disable the creation of the bridge entirely.

Together with Bridgeless Mode, we're also shipping a Native Module Interop Layer, that will allow you to reuse your old modules when in Bridgeless Mode. The [Renderer Interop Layer](https://github.com/reactwg/react-native-new-architecture/discussions/135) introduced in React Native 0.72 has also been adapted to work with Bridgeless Mode. 

As the rest of the New Architecture, Bridgeless Mode is considered experimental. We invite interested users to enable it and report to us problems and incompatibilities you might face in the New Architecture working group. There you can [read more about this in  details and report your feedback](TODO).
```

We've updated React Native's template on Android to use Kotlin instead of Java. The new `MainActivity.kt` and `MainApplication.kt` files are 36% smaller in size.

The [Upgrade Helper](https://react-native-community.github.io/upgrade-helper/) has also been updated to make it easier to migrate your `.java` files to `.kt` files.
If you've previously modified the Java files in your project and you need support migrating them to Kotlin, you can use the `Code > Convert Java file to Kotlin File` utility of Android Studio (also accessible with the shortcut <kbd>Cmd ⌘</kbd> + <kbd>Shift ⇧</kbd> + <kbd>Option ⌥</kbd> + <kbd>K</kbd>).

### Android 14 Support

We've updated React Native to fully support Android 14. Starting from 0.73, React Native developers can now target the latest Android SDK version, [API Level 34](https://developer.android.com/guide/topics/manifest/uses-sdk-element?hl=en#ApiLevels) (_Upside Down Cake_).

> 

Reviewer: @huntie can you cross check and include?

> 

Reviewer 4: this is a big one, it should be included

We recommend using [Hermes](/docs/hermes) for a consistent debugging experience on all platforms.
:::

## Breaking Changes

### Babel Package Renames

We've relocated two Babel-related packages out of Metro and into React Native's repository and versioning scheme, enabling us to simplify maintenance and upgrades. The new versions of these packages support New Architecture features in 0.73, meaning these dependencies must be updated.

Please follow the [Upgrade Helper](https://react-native-community.github.io/upgrade-helper/) when upgrading, to ensure you have updated these dependencies. Some packages have been renamed:

| Old Package Name                       | New Package Name                        |
| -------------------------------------- | --------------------------------------- |
| `metro-react-native-babel-preset`      | `@react-native/babel-preset`            |
| `metro-react-native-babel-transformer` | `@react-native/metro-babel-transformer` |

:::info
`@react-native/babel-preset` now includes `@react-native/babel-plugin-codegen`, this no longer needs to be specified separately in your Babel config file.
:::

> ### Other Breaking Changes

Reviewer: We need to mention somehwere that Java 17 is a requirement. @huntie if possible I would cross reference the Android 14 paragraph
